lucene/misc/src/java/org/apache/lucene/misc/store/DirectIODirectory.java

```diff
@@ -74,12 +65,12 @@
  *
  * @lucene.experimental
  */
-public class NativeUnixDirectory extends FSDirectory {
```

Comment:
I think we should make a copy of the NativeUnixDirectory, modify this to direct IO, then perhaps benchmark how they perform? If we replace in-place we won't be able to do it (unless you compile from different git commits). Then any removal of native code, should it follow-up, would be a cleaner patch as well.

Reply:
I don't think any explicit perf testing is needed, as we just replace the calls to our CPP file by setting the flag on open of FileChannel.

It's good that I have seen this pull request, because I was wondering why we took the "make native builds available" at all. 

We should also remove WindowsDirectory, because this one also works on Windows.

Reply:
Ok then - up to you! It was a fun exercise to combine native code anyway. :)

Reply:
I am not fully sure if the ExtendedOpenOption works at all with Windows. When checking the OpenJDK sources it seems to throw UOE, but haven't tested with new JDKs yet.

Anyways, we should try to get rid of WindowsDirectory, too, as the intention behind it is the same: do direct IO, it's just named differently on Windows.

We should maybe also look into the other ExtendedOpenptions.

Reply:
Thanks @dweiss and @uschindler for the reviews! The reason I originally opened another independent PR to make native build, and replaced the `NativeUnixDirectory` code in place here with the new flag was that, when I played with the luceneutil tool a bit, I noticed it took two complete lucene-solr check outs and run them side-by-side for performance testing. Plus there's more code in native code than just open direct setting. So I thought it might work better here if these two changes happened independently.

Will push another commit to remove `WindowsDirectory` in a day, as it's getting too late for me now :)

Reply:
IMHO, we should add more Javadocs, that this directory may throw `UnsupportedOperationException` if the underlying platform does not allow direct IO.

Reply:
I can confirm, windows works: http://cr.openjdk.java.net/~bpb/8164900/webrev.19/

Reply:
No worries, Zach. I think that native build part was a fun exercise! Code comes and goes, it's normal. Sleep well.

Reply:
> Will push another commit to remove WindowsDirectory in a day, as it's getting too late for me now :)

We should make a completely separate PR out of this. WindowsDirectory is also native, but we should still discuss if it was ever used. On Windows, currently you can only use MMapDirectory in a highly concurrent environment, because of the locking problem with positional reads in FileChannel.

Reply:
> IMHO, we should add more Javadocs, that this directory may throw UnsupportedOperationException if the underlying platform does not allow direct IO.

Ops yes I saw saw as well, but forgot to add earlier. Just added more java docs in latest commit.

> No worries, Zach. I think that native build part was a fun exercise! Code comes and goes, it's normal. Sleep well.

Haha thanks Dawid! I've learnt a great deal from that exercise as well!

> On Windows, currently you can only use MMapDirectory in a highly concurrent environment, because of the locking problem with positional reads in FileChannel.

I would like to learn more about this issue but can't seems to find in java doc for `ExtendedOpenOption.DIRECT`. Is this the one you are referring to:
1. https://lucene.apache.org/core/8_1_1/core/org/apache/lucene/store/FSDirectory.html  "Unfortunately, due to a Windows-only Sun JRE bug this is a poor choice for Windows, but on all other platforms this is the preferred choice."
2. https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6265734

Reply:
You are right, WindowsDirectory is unrelated to direct IO. So let's discuss this on separate issue! The background is here: https://issues.apache.org/jira/browse/LUCENE-2791. The issue behind that is the following: https://bugs.openjdk.java.net/browse/JDK-6265734